### PR TITLE
Add Dynamic Type support for iOS (Paper and Fabric)

### PR DESCRIPTION
**Pull request by amgleitman**
## Summary

This adds Dynamic Type support in iOS as described [here](https://github.com/react-native-community/discussions-and-proposals/issues/519).

`Text` elements have a new prop, `dynamicTypeRamp`, that allows users to specify which font ramp a particular `Text` element should take on as the OS's accessibility setting for text size. The different types line up with different values of `UIFontTextStyle`. If not specified, we default to the current behavior.

~~For the moment, this change is only for Paper. I tried applying a corresponding change to Fabric by adding an additional field to [`facebook::react::TextAttributes`](https://github.com/facebook/react-native/blob/main/ReactCommon/react/renderer/attributedstring/TextAttributes.h) and changing [`RCTEffectiveFontSizeMultiplierFromTextAttributes`](https://github.com/facebook/react-native/blob/afb124dcf0cdf0db525acc7cfd2cea2742c64068/ReactCommon/react/renderer/textlayoutmanager/platform/ios/RCTAttributedTextUtils.mm#L79-L84) to use that new field, but in the process I discovered that this function doesn't seem to ever get called, hence [this bug](https://github.com/facebook/react-native/issues/34990).~~

## Changelog

[iOS] [Added] - Dynamic Type support

## Test Plan

Validated with a test page in RNTester. Screenshots follow:

A) Default text size
B) Largest non-accessibility text size
C) Largest accessibility text size, split across two screenshots due to size

| A | B | C |
|-|-|-|
| Simulator Screen Shot - iPad (9th generation) - 2022-10-18 at 16 17 08 | Simulator Screen Shot - iPad (9th generation) - 2022-10-18 at 16 17 55 | Simulator Screen Shot - iPad (9th generation) - 2022-10-18 at 16 18 33 |
||| Simulator Screen Shot - iPad (9th generation) - 2022-10-18 at 16 18 42 |

### CODE REVIEW

**Comment by NickGerleman** [comment]
Hey there. When we talked about this PR a bit internally, there was a little bit of concern on accepting new capabilities which are paper-only. Having functionality that works on Paper, but not Fabric, means we would be taking new features away when the new Architecture is rolled out/default in OSS.

I see that the Fabric implementation is blocked by a separate bug in #34990. I will make sure this is surfaced internally.

**Comment by amgleitman** [comment]
@NickGerleman Now that e9b89b5ff2b0085128bec2879d761352747d6714 is a thing, I've added a Fabric implementation as well.

File: ReactCommon/react/renderer/textlayoutmanager/platform/ios/RCTAttributedTextUtils.mm [hunk 1]
```diff
@@ -55,6 +55,77 @@ inline static UIFontWeight RCTUIFontWeightFromInteger(NSInteger fontWeight)
...
+      return UIFontTextStyleTitle1;
+    case DynamicTypeRamp::LargeTitle:
+      return UIFontTextStyleLargeTitle;
+  }
+}
+
+inline static CGFloat RCTBaseSizeForDynamicTypeRamp(const DynamicTypeRamp &dynamicTypeRamp) {
+  switch (dynamicTypeRamp) {
+    case DynamicTypeRamp::Caption2:
+      return 11.0;
```
Comment (NickGerleman) [inline ReactCommon/react/renderer/textlayoutmanager/platform/ios/RCTAttributedTextUtils.mm:88]: From @sammy-SC:
> where do these numbers come from? Can we document the origin?

File: Libraries/Text/RCTTextAttributes.h [hunk 2]
```diff
@@ -7,6 +7,7 @@
 
 #import <UIKit/UIKit.h>
 
+#import <React/RCTDynamicTypeRamp.h>
```
Comment (NickGerleman) [inline Libraries/Text/RCTTextAttributes.h:10]: Seeing this error on an internal build script:
> packages/rn-tester/Pods/Headers/Public/React-Core/React/RCTTextAttributes.h:10:9: fatal error: 'React/RCTDynamicTypeRamp.h' file not found
#import <React/RCTDynamicTypeRamp.h>
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.

`RCTTextDecorationLineType` lives in `React/Views`. Wondering if this header should go there as well, vs if there is some internal bits we need to update.

File: ReactCommon/react/renderer/textlayoutmanager/platform/ios/RCTAttributedTextUtils.mm (hunk 1 above)
Comment (amgleitman) [reply ReactCommon/react/renderer/textlayoutmanager/platform/ios/RCTAttributedTextUtils.mm:89]: A link has been added with 98bd7b061776ca2f26a9c616a3cd56d1840a1fe1.

File: Libraries/Text/RCTTextAttributes.h (hunk 2 above)
Comment (amgleitman) [reply Libraries/Text/RCTTextAttributes.h:10]: Not sure what `RCTTextDecorationLineType` has to do with any of this. Considering that RCTDynamicTypeRamp.h is a newly created file with this PR, I'm guessing 0254c3a should help here.

**Comment by NickGerleman** [comment]
Seeing the below when we are invoking the Cocoapods build internally via `react-native-oss-ios-jsc`, but not coverage in CircleCI. @cipolleschi @hramos any guesses?

> The following build commands failed:
>>	CompileC <>/Library/Developer/Xcode/DerivedData/RNTesterPods-avhcctdswdycllfjzyasqkbrubnn/Build/Intermediates.noindex/Pods.build/Debug-iphonesimulator/React-RCTText.build/Objects-normal/x86_64/RCTUITextView.o <>/Libraries/Text/TextInput/Multiline/RCTUITextView.m normal x86_64 objective-c com.apple.compilers.llvm.clang.1_0.compiler (in target 'React-RCTText' from project 'Pods')
(1 failure)

**Pull request closed without merging**